### ISSUE TITLE
Fix duplicate trigger for tumor normal samples in back-to-back Runs

### ISSUE THREAD

**Issue by victorskl** [umccr/data-portal-apis#475]
**Context:**

In our WGS scenario; there can be such that 

- a Subject has tumor and normal samples
- a tumor sample is sequenced with Run 106
- a normal sample is sequenced with Run 216
- both Sequencing Runs 106 and 216 is back-to-back Runs (UPDATED: also affected 109/110)
- related https://umccr.slack.com/archives/C94CQMNVA/p1656832202406229

**Side Effects:**

- Current tumor normal automation is yet to support this scenario.
- Pipeline trigger twice for T/N workflows as it has seen 2 times during the Subject's samples pairing from each run. i.e. without aware of each other.
- Data overwrite is protected by `portal_run_id`, however though. Hence, only duplicate results.

**Actions:**

- Investigate and fix it

**Comment by victorskl** [comment]
I still feel that, as far as pipeline automation-wise;

Every workflow run is somehow induced by a SequenceRun i.e. "Per Run Context". Hence, there should be a link between Workflow and Sequence. If this link is broken, one piece of information has lost; automation-wise.

Whether or not, a workflow inputs need to find and pair tumor/nornal go across over multiple sequencing is another separation of concern.

One to tinker...

**Comment by victorskl** [comment]
Had another discussion with @reisingerf 

For now, we shall solve it with "sample's pre-step state" i.e. 

> **In order to stage a TN pair for a workflow run, all participating tumor and normal libraries' `wgs_alignment_qc` workflow run state must be succeeded.**

This would solve it. However. Note few caveats.
1) Overlapping Window -- the overlapping window should be wide enough. Usually, this is hour(s) apart based on previous runs' observation.
2) Tight Dependency -- auto TN paring now having tighter dependency on its pre-step workflow state; meaning, if there is no `wgs_alignment_qc` done, auto pairing would return none. This shall be desirable in most cases; business logic-wise.

@ohofmann @alexiswl Please feedback, if any.

**Comment by alexiswl** [comment]
This isn't going to be a trivial solution -

My solution would be to filter in the jobs in this final step [here](https://github.com/umccr/data-portal-apis/blob/dev/data_processors/pipeline/orchestration/tumor_normal_step.py#L219-L221).  

As far as I can see a minimal example of this issue is as follows:
* Run A contains tumor sample X
* Run B contains normal sample Y
* Fastqs are registered into database at completion of BCLConvert.
* But delay in tumor-normal processing after BCLConvert (because of wgs_alignment_qc) means both orchestration of Run A and Run B can see the fastq_list_row objects of the other run before creation of tumor normal workflow meaning both Run A and Run B both 'think' the previous runs came through beforehand.  

I think the Tight Dependency option @victorskl makes a lot sense here. And would probably implement this right at the end [here](https://github.com/umccr/data-portal-apis/blob/5851a0df8c6ea14a929ad926513689157b9896ef/data_processors/pipeline/orchestration/tumor_normal_step.py#L219-L221) as to not mess with existing logic but instead just filter out these particular edge cases.

Alternatively we update the `get_fastq_list_row_by_sequence_name` method [here](https://github.com/umccr/data-portal-apis/blob/5851a0df8c6ea14a929ad926513689157b9896ef/data_processors/pipeline/services/fastq_srv.py#L67-L74) which restricts libraries to those that have passed the wgs alignment qc process. 

Both options need to consider samples on adjacent runs being split over multiple lanes (similar to how we do it with tn runs [here](https://github.com/umccr/data-portal-apis/blob/5851a0df8c6ea14a929ad926513689157b9896ef/data_processors/pipeline/orchestration/tumor_normal_step.py#L33-L40). I think it makes sense to only consider libraries on adjacent sequence runs where all wgs alignment runs are complete on that sequence run.

**Comment by reisingerf** [comment]
Thanks for the update @alexiswl 

Victor and I had very similar ideas, basically only retrieve FASTQs if they are OK to be used (i.e. the QC has passed). We were thinking about adding that bit of logic around [here](https://github.com/umccr/data-portal-apis/blob/5851a0df8c6ea14a929ad926513689157b9896ef/data_processors/pipeline/orchestration/tumor_normal_step.py#L171) to avoid any further processing early on, but happy to look into your suggestion.

With regards to updating `get_fastq_list_row_by_sequence_name`, I'm not sure I like that as much, as it means adding additional semantics to a method that's now free of it. I.e. it would have to take into account cases where we don't have a QC workflow and now I can retrieve FASTQs independent of QC status (which may be a valid use case).

> need to consider samples on adjacent runs being split over multiple lanes
Is that not already covered?

**Comment by alexiswl** [comment]
> Why that limitation?

This is an issue with the English language and where I placed the word 'only' makes things ambiguous.

Let me re-phrase

_~~only~~ consider libraries on adjacent sequence runs **only** where all wgs alignment runs are complete_

> Is that not already covered?

I'm not sure. Given that we WGS per lane, there will be multiple wgs alignment runs for a sample that don't all finish at the same time. If we went with a QC passed method then we may pick up only a portion of the completed lanes on the adjacent run. 

> To cover for cases where a failed QC for some samples indicates issues for the whole run?

Discussed in point 2 above - not so much for failed QC, but I'd say we're better off waiting for an entire wgs alignment 'batch' to finish before considering any of those libraries 'available' to use.

**Comment by victorskl** [comment]
> Given that we WGS per lane, there will be multiple wgs alignment runs for a sample that don't all finish at the same time. If we went with a QC passed method then we may pick up only a portion of the completed lanes on the adjacent run.

Argh! Lane. Ok. My brain hurt! 😅 

Hold on, if we pool `wgs_alignment_qc` from (Subject) LibraryRun * ---> * Workflow [link](https://github.com/umccr/data-portal-apis/blob/dev/docs/model/data_portal_model.pdf), and check their status, should gives us "a world view" at that point. Right? 

LibraryRun at lane level, and they were populated as early as a sequencing is started.

Probably, we might wait you. When you are back then let catchup on this. Meanwhile let us take one step back & tinker a bit more.

Bottom line, we gotta make a naive lock -- to make them wait all, if there detects this pattern -- i.e. multiple sequencing runs within specified time window. But that's worst case scenario.

**Comment by reisingerf** [comment]
Hm, as @victorskl said, perhaps better to discuss/brainstorm/whiteboard in a chat... 

I think we'd be fine to check if a library (including all its, at that time known, LibraryRuns) have passed QC? 

As long as any LibraryRun for a given library is still running QC (or has failed), it would abort. That would include the same library on other lanes of the same run or topups and reruns across runs (as long as they are already known). Only once the "library" as a whole (i.e. all FASTQs that have been generated) have passed QC, the next steps would be taken.

There would be a very small window where two LibraryRuns could finish at the same time and both see the other as having passed QC and therefore both would kick off the next step, but that's a (hopefully) very rare race condition we'd have to deal with separately.


Ultimately, I'd like to get to a scenario where the "state" of a library/sample/patient is recorded and can be queried, independent of the workflows that led to said state. E.g. I'd like to separate out the logic whether or not a library is ready to be used in a workflow from the actual workflow processing/prep.

**Comment by victorskl** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Fix T/N pairing logic to filter out failed runs](https://trello.com/c/Fi2M4e68/1206-fix-t-n-pairing-logic-to-filter-out-failed-runs)
